### ISSUE TITLE
A better defined serial protocol for interacting with computer

### ISSUE THREAD

**Issue by eras** [UnifiedEngineering/T-962-improvements#22]
It would be nice if there was a better defined protocol for interacting with the computer. A simple one could for example prefix all current messages with some identifier indicating that they are free-form (a TUI/GUI could just pass those on to the user) but in addition have some messages that can be easily parsed and produced in a TUI/GUI/CLI tool. For example, "M Starting" or "P0", or "M Temperature 190.42 degrees", "P1 190.42".

Configuring profiles and monitoring the process could be much nicer from a work station.

**Comment by jieter** [comment]
I've code a simple serial protocol in #51 (including plotting with python/matplotlib), but did not pay attention to this discussion. Any thoughts @eras @colinoflynn?

**Comment by colinoflynn** [comment]
code > talk ;-)

I'd seen progress being made, and figured if you've got stuff working that's the best option. Can always hack onto it later if required.

**Comment by jieter** [comment]
Now that #51 is merged, lets do some talking. What kind of functionality should be added to the serial interface? Some thoughts:

- Bake should get timer functionality
- Dump profile to serial
- Update/load a profile

Besides that, replies of the oven are not really standardized now, sow polishing that would be another thing to consider doing.

**Comment by xnk** [comment]
The commands works well from the computer but has to be sent as one continuous stream, typing them in manually doesn't work as expected. Not a major issue when using the python wrapper but if talking to it using a terminal program you have to basically paste in what you want to send. Timer on bake is nice. That timer should ideally start ticking only when having reached the setpoint. This will make it possible to use this functionality as a pre-heat for the oven as well.

**Comment by jieter** [comment]
Good point, I tested with a program called `cutecom` which only sends the command after you press enter.

About the timer in bake function: If I pre-heat, the reflow profile will try to cool down the oven before heating it again. So pre-heat only makes sense if the profile starts with a relatively high temperature?

**Comment by xnk** [comment]
Aha, makes sense! I used my favorite terminal program "screen"!

Yes for pre-heat you don't want to heat the oven too much anyway, but letting the oven reach 50C and then staying there for a few minutes definitely makes a difference if the oven was very cold to begin with. The reflow profiles were all adjusted to start at 50C some time ago for the same reason, to avoid reflow starting to cool right from the start..
